### PR TITLE
Google-Signup-functionality-#46

### LINKED ISSUE #46: Implement Google Signup Functionality
Users should be able to sign up using their Google tokens for authentication.

This feature enhances user convenience by allowing them to sign up using their Google accounts, reducing the need to remember additional login credentials.

Screenshots
Actual:
image

Expected :
image

Could you please Assign this issue to me under gssoc

### PR DESCRIPTION
**Pull request by mr0nerd**
# Pull Request

### Title

Add Google Sign-In and Sign-Up Functionality via Firebase
### Description

This pull request introduces Google sign-in and sign-up functionality using Firebase for authentication. The frontend component handles the Google sign-in process, and the backend controller manages user creation and authentication.
### Related Issues

Fixes #46 
### Changes Made

- Added Google OAuth component to the frontend for sign-in/sign-up using Firebase
- Created a `googleAuthController` 
- Generated a random password for new Google users
- Validated user data using `signupBodySchema`
- Created tokens for authenticated users
- Sent a welcome email to newly created users
### Checklist 

- [x] I have tested the changes locally
- [ ] Documentation has been updated (if necessary)
- [x] Changes are backward-compatible

### Screenshots (if applicable)

Screenshot 2024-07-15 171936

Modified SignUp/SignIn pages

Screenshot 2024-07-15 172033

Using "Continue with Google" button to SignUp or SignIn

Screenshot 2024-07-15 171913

New user is created if its not already existing 

Screenshot 2024-07-15 171958

Welcome mail is sent to new user 

Screenshot 2024-07-15 172014

### Additional Notes

This PR enhances the authentication system by allowing users to sign in and sign up using their Google accounts via Firebase, ensuring a smoother and faster onboarding process.
Footer

### CODE REVIEW

File: frontend/src/firebase.ts
Comment (VaibhavArora314) [inline frontend/src/firebase.ts:1]: Convert these credentials to secrets and can you also find a way to pass environment variables while serving this through backend as currently in the deployment, backend provides react app as a static resource

**Review by VaibhavArora314 — CHANGES_REQUESTED** [review]

**Comment by akbatra567** [comment]
Hey @VaibhavArora314, 
We have another open PR for 
https://github.com/VaibhavArora314/StyleShare/pull/358

File: frontend/src/firebase.ts
Comment (mr0nerd) [reply frontend/src/firebase.ts:1]: I have converted the firebase credentials to secret.

Now to pass environment variables, i am doing following changes in the app.ts file of backend please review it 

import express, { NextFunction, Request, Response } from 'express';
import rootRouter from './routes';
import bodyParser from 'body-parser';
import cors from 'cors';
import path from 'path';
import { config } from 'dotenv';
import fs from 'fs';

config();

const app = express();
app.use(cors());
app.use(bodyParser.json());

app.get('/', (req: Request, res: Response) => {
  res.redirect('/app');
});

// Serve frontend with environment variables
app.use('/app', express.static(path.join(__dirname, '../../frontend/dist'), { maxAge: 0 }));
app.get('/app/*', (req: Request, res: Response) => {
  const indexPath = path.join(__dirname, '../../frontend/dist', 'index.html');
  let indexContent = fs.readFileSync(indexPath, 'utf-8');

  indexContent = indexContent.replace(
    '</head>',
    `<script>
      window.env = {
        VITE_FIREBASE_API_KEY: "${process.env.VITE_FIREBASE_API_KEY}",
        VITE_FIREBASE_AUTH_DOMAIN: "${process.env.VITE_FIREBASE_AUTH_DOMAIN}",
        VITE_FIREBASE_PROJECT_ID: "${process.env.VITE_FIREBASE_PROJECT_ID}",
        VITE_FIREBASE_STORAGE_BUCKET: "${process.env.VITE_FIREBASE_STORAGE_BUCKET}",
        VITE_FIREBASE_MESSAGING_SENDER_ID: "${process.env.VITE_FIREBASE_MESSAGING_SENDER_ID}",
        VITE_FIREBASE_APP_ID: "${process.env.VITE_FIREBASE_APP_ID}"
      };
    </script></head>`
  );
  res.setHeader('Cache-Control', 'no-store, no-cache, must-revalidate, private');
  res.send(indexContent);
});

app.use('/admin', express.static(path.join(__dirname, '../../admin/dist'), { maxAge: 0 }));
app.get('/admin/*', (req: Request, res: Response) => {
  res.setHeader('Cache-Control', 'no-store, no-cache, must-revalidate, private');
  res.sendFile(path.join(__dirname, '../../admin/dist', 'index.html'));
});

app.use('/api/v1', rootRouter);

app.use((err: Error, req: Request, res: Response, next: NextFunction) => {
  console.error(err.stack);
  res.status(500).json({ error: 'Internal Server Error' });
});

export default app;

**Comment by VaibhavArora314** [comment]
> Hey @VaibhavArora314, We have another open PR for #358 Which might be on the same lines to this.

hey @akbatra567, I asked that person if she is able to work on this issue in #46 due to repeated same problems or want someone else to work on it and she was okay with someone else working on it. Sorry, I forgot to close that pull request causing confusion.

File: frontend/src/firebase.ts
Comment (VaibhavArora314) [reply frontend/src/firebase.ts:1]: if this works, do the changes in this pr itself so we can review it

**Comment by mr0nerd** [comment]
i have made the firebase config data secret ,but please review it.

**Review by VaibhavArora314 — CHANGES_REQUESTED** [review]
resolve merge conflicts, and have you also added logic to pass env var to frontend?

**Comment by mr0nerd** [comment]
All the requested changes have been made and merge conflicts has been resolved .so please review it.
Also, I am contributing to this repo as a GSSOC'24 contributor, so please add GSSOC  tag on this pr

**Review by VaibhavArora314 — CHANGES_REQUESTED** [review]
you are passing the env variables via an HTTP request currently which is a very big security loophole. is there no other way to pass the env variables securely to the frontend?

**Comment by VaibhavArora314** [comment]
On searching online, I can see that there are some ways where a token is sent from frontend to backend on google auth and it is verified on the backend then the user is signed in. Maybe you can use this method to avoid the problem of environment variables.

**Pull request closed without merging**